packages/back-end/src/services/experimentTimeSeries.ts

```diff
@@ -0,0 +1,294 @@
+      sourceId: experiment.id,
+      metricId,
+      lastExperimentSettingsHash: experimentHash,
+      lastMetricSettingsHash: getMetricSettingsHash(
+        metricId,
+        experimentSnapshot.settings.metricSettings.find(
+          (it) => it.id === metricId
+        )!,
+        factMetrics,
+        factTableMap
+      ),
+      singleDataPoint: {
+        date: experimentSnapshot.dateCreated,
+        variations: timeSeriesVariationsPerMetricId[metricId],
+      },
+      tags:
+        notificationsTriggered.length > 0 || hasSignificantDifference
+          ? ["triggered-alert"]
+          : undefined,
+    })
+  );
+
+  await context.models.metricTimeSeries.upsertMultipleSingleDataPoint(
+    metricTimeSeriesSingleDataPoints
+  );
+}
+
+function convertMetricToMetricValue(
+  metric: SnapshotMetric | undefined
+): MetricTimeSeriesValue | undefined {
+  if (!metric) {
+    return undefined;
+  }
+
+  // NB: Explicitly naming all fields to benefit from type safety
+  // when SnapshotMetric and MetricTimeSeriesDataPoint change
+  return {
+    value: metric.value,
+    denominator: metric.denominator,
+    expected: metric.expected,
+    ci: metric.ci,
+    pValue: metric.pValue,
+    pValueAdjusted: metric.pValueAdjusted,
+    chanceToWin: metric.chanceToWin,
+  };
+}
+
+const hashObject = (obj: object) => md5(JSON.stringify(obj));
+
+function getExperimentSettingsHash(
+  snapshotSettings: ExperimentSnapshotSettings,
+  snapshotAnalysisSettings: ExperimentSnapshotAnalysisSettings
+): string {
+  return hashObject({
+    // snapshotSettings
+    activationMetric: snapshotSettings.activationMetric,
+    attributionModel: snapshotSettings.attributionModel,
+    queryFilter: snapshotSettings.queryFilter,
+    segment: snapshotSettings.segment,
+    skipPartialData: snapshotSettings.skipPartialData,
+    datasourceId: snapshotSettings.datasourceId,
+    exposureQueryId: snapshotSettings.exposureQueryId,
+    startDate: snapshotSettings.startDate,
+    regressionAdjustmentEnabled: snapshotSettings.regressionAdjustmentEnabled,
+    experimentId: snapshotSettings.experimentId,
+
+    // analysisSettings
+    dimensions: snapshotAnalysisSettings.dimensions,
+    statsEngine: snapshotAnalysisSettings.statsEngine,
+    regressionAdjusted: snapshotAnalysisSettings.regressionAdjusted,
+    sequentialTesting: snapshotAnalysisSettings.sequentialTesting,
+    sequentialTestingTuningParameter:
+      snapshotAnalysisSettings.sequentialTestingTuningParameter,
+    baselineVariationIndex: snapshotAnalysisSettings.baselineVariationIndex,
+    pValueCorrection: snapshotAnalysisSettings.pValueCorrection,
+  });
+}
+
+function getMetricSettingsHash(
+  metricId: string,
+  metricSettings: MetricForSnapshot,
+  factMetrics?: FactMetricInterface[],
+  factTableMap?: Map<string, FactTableInterface>
+): string {
+  const factMetric = factMetrics?.find((metric) => metric.id === metricId);
+  if (!factMetric) {
+    return hashObject(metricSettings);
```

Comment:
We might want to have a fallback here in case the metricSettings object is undefined, as it can be if somehow the metric got deleted but is still in the experiment model.